feat (hset): Support arguments (count, withvalues) in HRANDFIELD

theyueli opened the pull request:
Fixes: https://github.com/dragonflydb/dragonfly/issues/858

It might possibly fix: https://github.com/dragonflydb/dragonfly/issues/1707

The algorithms that support both encodings (string map and listpack) have been implemented and tested. To use string map requires a larger hset (my tests used 500+ entries)

The random selection algorithms implemented for string map class are the reimplementation of the same algorithms used by Redis' listpack. (therefore same time complexities)

Without this patch:

```
127.0.0.1:6379> HSET coin heads obverse tails reverse edge null
(integer) 3
127.0.0.1:6379> HRANDFIELD coin 3 WITHVALUES
(error) ERR wrong number of arguments for 'hrandfield' command
127.0.0.1:6379> HRANDFIELD coin -1
(error) ERR wrong number of arguments for 'hrandfield' command
127.0.0.1:6379> 
```
after this patch:

```
127.0.0.1:6379> HSET coin heads obverse tails reverse edge null
(integer) 3
127.0.0.1:6379> HRANDFIELD coin -5 WITHVALUES
 1) "tails"
 2) "reverse"
 3) "tails"
 4) "reverse"
 5) "heads"
 6) "obverse"
 7) "tails"
 8) "reverse"
 9) "edge"
10) "null"
127.0.0.1:6379> HRANDFIELD coin 3
1) "heads"
2) "tails"
3) "edge"
127.0.0.1:6379> HRANDFIELD coin 3 WITHVALUES
1) "heads"
2) "obverse"
3) "tails"
4) "reverse"
5) "edge"
6) "null"
```

royjacobson commented:
redis claim `O(N)` complexities with `N` being the amount of items returned. AFAICT `RandomPairsUnique` and `RandomPairs` are `O(M)` with `M=size(map)`. Am I missing something?

dranikpg commented:
@royjacobson 

I might be wrong, but it seems like listpacks implementation is also O(M) 😮 They use the same algorithms as here

Actually we _can_ make our implementation O(N) by harnessing stringmaps internals. But I don't think its mandatory as I already answered in the first PR on this issue

dranikpg commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
   return GetValue(str);
 }
 
+void StringMap::RandomPair(sds* key, sds* value) {
+  auto itr = begin();
+  for (unsigned int i = 0; i < (rand() % Size()); ++i)
+    ++itr;
+  *key = itr->first;
+  if (value != nullptr)
+    *value = itr->second;
+  return;
+}
```
I'm not a fan of C style inside out C++ codebase. Lets return those pointers by value

```c++
pair<sds, sds> RandomPair() {
  auto it = begin();
  std::advane(it, rand % size);
  return it;
}

```

theyueli replied:
very nice, thanks!

theyueli replied:
actually more work needs to be done as StringMap's iterator doesn't support std::advance.

theyueli replied:
@dranikpg i just moved the logic into an += operator, see if that looks ok?

kostasrim replied:
You can also add the `difference_type` alias in the Iterator. It's an one liner :smile:

dranikpg commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
+  return;
+}
+
+unsigned int StringMap::RandomPairsUnique(unsigned int count, sds* keys, sds* vals) {
+  unsigned int total_size = Size();
```
Same here, lets return a vector<pair<sds,sds>>. And then you don't need create raw arrays manually or the repetitive `if (!with_vales.empty()) clause`

theyueli replied:
@dranikpg I prefer only return vals upon request. The count value theoretically can be large, and the value itself could be large. Unlike RandomPair which only returns one pair, this function potentially could significantly increase memory consumption if we always return everything.

theyueli replied:
the function now takes in two vecotor references (keys, vals), and a boolean value indicating whether the `vals` needs to be filled in.

kostasrim replied:
Then just return a std::pairt<vector, vector>, the second one will be empty and therefore no memory will wasted for an empty vector other than the pointer + size in the second field of pair

theyueli replied:
@kostasrim please take a look at the current implementation. I think it is a personal choice to use two vectors separately or a pair of vectors...

kostasrim replied:
I think it doesn't make much sense to have the two vectors as input parameters. If you do what @dranikpg suggested and return them instead, we could squash that function into one. Furthermore, I see no reason why from an API pov I would want to pass the result vectors as input parameters for this particular contexts. It's just noise.

dranikpg commented on src/server/hset_family.cc:
```diff
@@ -1005,8 +1006,37 @@ void HSetFamily::HStrLen(CmdArgList args, ConnectionContext* cntx) {
+    }
+  }
+
+  if (args.size() == 3) {
+    ToUpper(&args[2]);
+    with_values = ArgS(args, 2);
+    if (with_values != "WITHVALUES")
+      return (*cntx)->SendError(kSyntaxErr);
+  }
```
That is a _very_ unusual way to set a boolean flag

just do `if (ArgS(args, 2) != "W...S") return error; with_values = true;`

theyueli replied:
`with_values` here is not a boolean flag, but i see what you mean. Will change it to a boolean value.

dranikpg commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1050,90 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
+        if (!with_values.empty())
+          vals = new sds[actual_count];
+        if (count >= 0) {
+          if (!with_values.empty()) {
+            string_map->RandomPairsUnique(actual_count, keys, vals);
+          } else {
+            string_map->RandomPairsUnique(actual_count, keys, nullptr);
```
1. It's already null, why the if
2. Lets don't use raw arrays and don't pass nullptr. The performance isn't gonna measurably drop if we return the values as well

theyueli replied:
> 1. It's already null, why the if

I don't understand this comment, you mean `vals` is already null? vals will be null if `WITHVALUES` is not specified. Otherwise, vals will be allocated through:

`
        if (!with_values.empty())
          vals = new sds[actual_count];
`

> 2\. Lets don't use raw arrays and don't pass nullptr. The performance isn't gonna measurably drop if we return the values as well

It was not for performance, it was a way to allow us fill vals only when `WITHVALUES` is specified. But I will change to use vector and pass in another boolean value with_values to achieve the same purpose.

kostasrim replied:
You can always return an empty vector if the `WITHVALUES` is not specified

theyueli replied:
yes, but now this `with_values` has been changed to a boolean value, it is much more clear and readable. Thanks!

dranikpg commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1050,90 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
+        // count has been specified.
+        if (count >= 0) {
+          // always returns unique entries.
+          if (!with_values.empty()) {
+            // WITHVALUES is specified.
+            // randomly returns actual_count unique keys and values.
+            lpRandomPairsUnique(lp, actual_count, keys, vals);
+          } else {
+            lpRandomPairsUnique(lp, actual_count, keys, NULL);
```
same here, its already null, so no if is needed

kostasrim replied:
Nice catch! This makes my suggestion with the `body` lambda even more plausible/simple. +1

theyueli replied:
@dranikpg  see my comments/questions above.

kostasrim commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
+  std::vector<RandomPick> picks;
+  unsigned int total_size = Size();
+
+  for (unsigned int i = 0; i < count; ++i) {
```
you can use a `std::map<uint32_t, uint32_t>`.  The key represents the index, and value represents the number of times you encountered it (we need map here, to since it's an ordered container). Each time you loop here, you check if the index `rand() % total_size` already exists. If it does you increment it's occurrence (the value) by 1. Otherwise (if the index is not already inserted) you insert it with an associated value of `1` (since it's the first occurrence).  

That way you will:

1. Get rid of `std::sort`.
2. Simplify the code below, since now you don't need two while loops. You only need a for loop, and for each element, you output `occurrence` number of times

dranikpg replied:
Good suggestion, but I don't think it matters that much. Sort only gets slow for a really big number of elements and for it the i/o time is much larger

kostasrim replied:
No no, I do not care about the performance here, I don't expect that `std::sort` will have any measurable impact since it the count will always be relatively small. But by using a `map` it will simplify the implemetnaion of this function and make it more readable

theyueli replied:
I prefer to keep as is. sort() is more implicit when std::map is used. Personally, the current version has better readabliy by making sort obvious.

kostasrim replied:
So a) I think stl associative ordered containers are used exactly for that, so I would argue that vector + sort is an antipattern. b) how is what you wrote simpler than this 3 lines? No pair accesses, no nested while loops no nothing:

```
for(auto it = begin(); it < end() && picks_sz < count ; ++it) {
  auto [key, frequency] = *it;
  keys.insert(keys.end(), frequency, key);
  picks_sz += frequency;
}
```

I might a have a mistake on the above code but that's the gist of it

kostasrim commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1050,90 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
+        if (!with_values.empty())
+          vals = new listpackEntry[actual_count];
+
+        // count has been specified.
```
Ignore me if you don't like this comment, but I think it will reduce the verbosity by a large factor. 

Notice that both of the branches use the exact same code only to call different functions. Therefore:

```
auto body = [](auto f_a, auto f_b, auto f_c, auto f_d, auto keys, auto vals) {
        if (count >= 0) {
          // always returns unique entries.
          if (!with_values.empty()) {
            // WITHVALUES is specified.
            // randomly returns actual_count unique keys and values.
            lpRandomPairsUnique(lp, actual_count, keys, vals);
          } else {
            lpRandomPairsUnique(lp, actual_count, keys, NULL);
          }
        } else {
          // allows non-unique entries.
          if (!with_values.empty()) {
            // WITHVALUES is specified.
            // randomly returns actual_count keys and values.
            lpRandomPairs(lp, actual_count, keys, vals);
          } else {
            lpRandomPairs(lp, actual_count, keys, NULL);
          }
        }

        for (size_t i = 0; i < actual_count; ++i) {
          StrVecEmplaceBack(str_vec, keys[i]);
          if (!with_values.empty()) {
            StrVecEmplaceBack(str_vec, vals[i]);
          }
        }
}
```

and now: 

```
auto f_a = [&]() { lpRandomPairsUnique(lp, actual_count, keys, vals) }
auto f_b = ...
auto f_c = ...
auto f_d = ...

finally body(f_a, f_b, f_c, f_d, keys, vals)
```

Slightly an overkill but it really reduces the duplication here :)

dranikpg replied:
A little overkill 🙂

kostasrim reviewed (commented):
Good work :man_cook: , some minor comments

dranikpg commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1050,90 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
+        size_t actual_count =
+            (count >= 0) ? std::min(size_t(count), string_map->Size()) : abs(count);
+        sds *keys = nullptr, *vals = nullptr;
+        keys = new sds[actual_count];
```
Using a vector here has no downsides, but I already suggested just returning a single vector

kostasrim replied:
Yes + 1

theyueli replied:
I will change to use vector for boh keys and vals, but I will still pass them by reference separately. vals will continue being filled only if `WITHVALUES` is specified.

theyueli commented on src/server/hset_family.cc:
```diff
@@ -1005,8 +1006,37 @@ void HSetFamily::HStrLen(CmdArgList args, ConnectionContext* cntx) {
 }
 
+void StrVecEmplaceBack(StringVec& str_vec, const listpackEntry& lp) {
+  if (lp.sval) {
+    str_vec.emplace_back(reinterpret_cast<char*>(lp.sval), lp.slen);
+  } else {
+    str_vec.emplace_back(absl::StrCat(lp.lval));
+  }
+  return;
```
fixed.

theyueli commented on src/server/hset_family.cc:
```diff
@@ -1005,8 +1006,37 @@ void HSetFamily::HStrLen(CmdArgList args, ConnectionContext* cntx) {
+  int32_t count;
+  string_view with_values;
+
+  if (args.size() > 1) {
+    if (!SimpleAtoi(ArgS(args, 1), &count)) {
```
fixed. 

regarding error on `count = 0`, this will make dragonfly behave differently from Redis. Redis still returns empty array in this case...

kostasrim replied:
> regarding error on count = 0, this will make dragonfly behave differently from Redis. Redis still returns empty array in this case...

Glad you looked, it's fine to keep it as is :)

theyueli commented on src/server/hset_family.cc:
```diff
@@ -1005,8 +1006,37 @@ void HSetFamily::HStrLen(CmdArgList args, ConnectionContext* cntx) {
+    }
+  }
+
+  if (args.size() == 3) {
```
this has been done at the beginning of this function:

```
  if (args.size() > 3) {
    DVLOG(1) << "Wrong number of command arguments: " << args.size();
    return (*cntx)->SendError(kSyntaxErr);
  }
```

kostasrim replied:
oh yes sorry!

theyueli commented on src/core/string_map.h:
```diff
@@ -117,6 +117,19 @@ class StringMap : public DenseSet {
+  // Randomly selects count of key value pairs. The selections are unique.
+  // if count is larger than the total number of key value pairs, returns
+  // every pair.
+  unsigned int RandomPairsUnique(unsigned int count, sds* keys, sds* vals);
```
changed

theyueli commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
+  return;
+}
+
+unsigned int StringMap::RandomPairsUnique(unsigned int count, sds* keys, sds* vals) {
```
the function might have other use cases in the future... so I will use your other suggestion to make it uint32_t

kostasrim replied:
There will be no other use cases for this function as the API is pretty stable. If it does change then we will change it accordingly. Otherwise, each time I have to look on that code I need to wonder why the return value is ignored. 

Also maybe just for our sanity, you can add a `DCHECK` that the size of the result vectors are the size of count (although I checked your algorithm -- just to be on the safe side -- extra checks don't do harm and are a safery barrier)

theyueli replied:
> Also maybe just for our sanity, you can add a `DCHECK` that the size of the result vectors are the size of count (although I checked your algorithm -- just to be on the safe side -- extra checks don't do harm and are a safery barrier)

Currently in RandomPairs, I have the following code

```
  keys.reserve(count);
  if (with_value)
    vals.reserve(count);
```

This sort of enforced the size constraint. 

But I will add the DCHECK in RandomPairsUnique.

theyueli replied:
> There will be no other use cases for this function as the API is pretty stable. If it does change then we will change it accordingly. Otherwise, each time I have to look on that code I need to wonder why the return value is ignored.

changed.

theyueli commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
+}
+
+void StringMap::RandomPairs(unsigned int count, sds* keys, sds* vals) {
+  typedef std::pair<unsigned int, unsigned int> RandomPick;
```
fixed

theyueli commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
+
+  sds key, val;
+  while (itr != end() && pick_index < count) {
+    key = itr->first;
+    val = itr->second;
```
fixed

theyueli commented on src/core/string_map.cc:
```diff
@@ -101,6 +101,74 @@ sds StringMap::Find(std::string_view key) {
   return GetValue(str);
 }
 
+void StringMap::RandomPair(sds* key, sds* value) {
```
correct, but i would like to keep it as a utility function for the class so that is independent from Redis' code.

kostasrim replied:
Then there is no point for duplication. Plz remove this function, it's noise.

theyueli replied:
> Then there is no point for duplication. Plz remove this function, it's noise.

I prefer keeping it as is. This is a very convenient function such that I do not need to pass in two vectors to `RandomPairsUnique` just to get one key value pairs.

theyueli commented:
@dranikpg @kostasrim addressed all the comments, please take another look, thanks!

theyueli commented:
> @dranikpg @kostasrim addressed all the comments, please take another look, thanks!

@romange it seems current main branch's CI is broken...

The following tests failed on the main branch without this patch.
50 - rdb_test (Failed)
61 - cluster_family_test (Failed)

dranikpg commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1048,73 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
+        size_t actual_count = (count >= 0) ? std::min(size_t(count), hlen) : abs(count);
+        listpackEntry *keys = nullptr, *vals = nullptr;
+        keys = new listpackEntry[actual_count];
+        if (with_values)
```
Better use a unique_ptr<listpackEntry[]>

Or vector, but calling data for empty ones might be problematic

theyueli replied:
fixed.

dranikpg commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1048,73 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
+          }
+        } else {
+          // allows non-unique entries.
+          if (with_values) {
```
You don't need this if

theyueli replied:
agreed, thanks!

dranikpg reviewed (dismissed):
1. But I still encourage you to use a unique_ptr instead of raw arrays
2. about RandomPairs(), we usually take output arguments by pointer (and returning values is even better) but I don't wanna be picky

kostasrim commented on src/server/hset_family.cc:
```diff
@@ -1020,24 +1048,61 @@ void HSetFamily::HRandField(CmdArgList args, ConnectionContext* cntx) {
-        str_vec.emplace_back(absl::StrCat(key.lval));
+        size_t actual_count = (count >= 0) ? std::min(size_t(count), hlen) : abs(count);
+        listpackEntry *keys = nullptr, *vals = nullptr;
+        keys = new listpackEntry[actual_count];
```
Oh wait, I did not see this. Plz use `absl::Cleanup`

kostasrim replied:
You delete this later on with `[]`. This is no RAII and raw memory managements and it's looking for trouble

theyueli replied:
I will use unique_ptr

theyueli commented:
> 1. But I still encourage you to use a unique_ptr instead of raw arrays

thanks @dranikpg, I just changed to use `unique_ptr`, please take a second look.

dranikpg reviewed (approved):
👍🏻
